### ISSUE TITLE
[CT-971] [SPIKE] Materialize multiple table for python model table materialization

### ISSUE THREAD

**Issue by ChenyuLInx** [dbt-labs/dbt-core#5595]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

With python model being added, we anticipate users would use python model to build out dataset for training and testing machine learning model. Instead of holding on to the previous one table per model approach, we should explore the option of materializing multiple table in one python model. This feature is only for python models and not intended to port to SQL models.

## Use Case description:
``` python
# dataset.py
def model(dbt, session):
  train_x, train_y, test_x, test_y = some_python_package.split_data(dbt.ref("raw_data"))
  return train_x, train_y, test_x, test_y
```
Here `train_x`, `train_y`, `test_x`, `test_y` are all pandas dataframes or other type of dataframe that specific warehouse vendor supports. We would want to create 4 tables from this model, and being able to use them in down stream models.

``` sql
select * from {{ref("dataset", "train_x")}}
``` 

## Areas to look into
- Getting the name of each table:
We would need user to return names variables in the return like the above example, `return some_python_package.split_data(dbt.ref("raw_data"))` would not be allowed. [This](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/parser/models.py#L41) is where we currently enforce rules we have for model definition. [This](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/parser/models.py#L110) is where we parse out info from files. One thing to figure out here is how to parse out the names from return. The reason that check is separate from actual visit is because `visit_Call` seems will not be run if we add `visit_FunctionDef` to the same visitor class

- Pass the parsed information to materialization logic
Look into how to properly get the table names to materilazation logic. Also need to check where we have the 1 model to 1 table assumption.(That might be the reason `source` and `ref` are separate commands)

- store results of run properly
look into what results we should return to properly store the model run results for dbt's cached relation to work properly. Currently [store_results](https://github.com/dbt-labs/dbt-core/blob/a7ff003d4f74763f126e77a0fc9c55e7a3311b3d/core/dbt/context/providers.py#L722) is being used in macro `statement` to store run results, then it will somehow update [RelationsCache](https://github.com/dbt-labs/dbt-core/blob/a7ff003d4f74763f126e77a0fc9c55e7a3311b3d/core/dbt/adapters/base/impl.py#L184) in the Adapter so that the following model build will be run correctly. we need to have that multi table python model properly populate that cache after run is done.

- being able to reference it from downstream models and not breaking existing functionality.
Our ref should work with something like this, checkout [source](https://docs.getdbt.com/docs/building-a-dbt-project/using-sources). Look into the logic to make it happen over there and determine whether it can be used for models

- dbt-docs
Can we do something similar to what source is doing also for models in order to make dbt-docs working as expected?

## Outcome of the ticket
All above mentioned uncertainties answered, and also we should be able to answer the question of: does the above mentioned items get us a workable version of multi table materialization for python model, what else are we missing. Document the findings under this SPIKE.

If the answer to last question is yes, create issues in `dbt-core` and adapters for the implementation of multi table materialization

### Describe alternatives you've considered

Create multiple models to materialize multiple tables. but this is not ideal since the computation for the described usecase will be duplicated, and sometimes that process is random so we can't do it in multiple models.

**Comment by lostmygithubaccount** [comment]
we should look to partner with an external customer/user or internal stakeholders to develop an actual data science project with these needs -- my further suggested acceptance criteria:

- a working example of this in production
- a devblog post -- either by us or a partner we can share -- to demonstrate the example and discuss pros/cons

as we've found out, a potentially simpler solution than supporting multiple outputs is simple adding a "label" column -- is this row in the train set, validation set, test set, etc?
